File: Medium/701 - InsertIntoABinarySearchTree/Solution.cs
/// Question 701. Insert into a Binary Search Tree (https://leetcode-cn.com/problems/insert-into-a-binary-search-tree/)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution
{
    // 执行用时: 160 ms, 在所有 C# 提交中击败了 77.78% 的用户
    // 内存消耗: 37.1 MB, 在所有 C# 提交中击败了 100.00% 的用户
    public TreeNode InsertIntoBST(TreeNode root, int val)
    {
        if (root == null) { return new TreeNode(val); }

        this.InsertIntoBinarySearchTree(root, val);
        return root;
    }

    public void InsertIntoBinarySearchTree(TreeNode root, int val)
    {
        if (root.val > val)
        {
            if (root.left == null)
            {
                root.left = new TreeNode(val);
                return;
            }
            else
            {
                InsertIntoBST(root.left, val);
            }
        }
        else
        {
            if (root.right == null)
            {
                root.right = new TreeNode(val);
                return;
            }
            else
            {
                InsertIntoBST(root.right, val);
            }
        }
    }
}


public class SolutionWithIteration
{
    // 执行用时: 164 ms, 在所有 C# 提交中击败了 61.11% 的用户
    // 内存消耗: 37 MB, 在所有 C# 提交中击败了 100.00% 的用户
    public TreeNode InsertIntoBST(TreeNode root, int val)
    {
        TreeNode newNode = new TreeNode(val);
        if (root == null) { return newNode; }

        TreeNode currentNode = root;
        while (true)
        {
            if (currentNode.val > val)
            {
                if (currentNode.left != null)
                {
                    currentNode = currentNode.left;
                }
                else
                {
                    currentNode.left = newNode;
                    return root;
                }
            }
            else
            {
                if (currentNode.right != null)
                {
                    currentNode = currentNode.right;
                }
                else
                {
                    currentNode.right = newNode;
                    return root;
                }
            }
        }
    }
}

File: Medium/701 - InsertIntoABinarySearchTree/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

/**
 * 执行用时：144 ms, 在所有 typescript 提交中击败了 52.00% 的用户
 * 内存消耗：45.7 MB, 在所有 typescript 提交中击败了 11.76% 的用户
 *
 * @param {(TreeNode | null)} root
 * @param {number} val
 * @returns {(TreeNode | null)}
 */
function insertIntoBST(root: TreeNode | null, val: number): TreeNode | null {
  if (!root) {
    return new TreeNode(val);
  }

  insertIntoTree(root, val);

  return root;
}

function insertIntoTree(root: TreeNode, val: number): void {
  if (root.val > val) {
    if (!root.left) {
      root.left = new TreeNode(val);
    } else {
      insertIntoTree(root.left, val);
    }
  } else {
    // root.val < val
    if (!root.right) {
      root.right = new TreeNode(val);
    } else {
      insertIntoTree(root.right, val);
    }
  }
}
